etc/eslint/rules/typescript.js

```diff
@@ -0,0 +1,2526 @@
+*
+* @example
+* // Bad...
+* class Bad extends Base {
+*   constructor() {
+*     console.log('missing super()');
+*   }
+* }
+*
+* // Good...
+* class Good extends Base {
+*   constructor() {
+*     super();
+*     console.log('has super()');
+*   }
+* }
+*/
+rules[ 'constructor-super' ] = 'error';
+
+/**
+* Enforces consistent use of curly brace conventions.
+*
+* @name curly
+* @memberof rules
+* @type {string}
+* @default 'error'
+* @see [curly]{@link https://eslint.org/docs/rules/curly}
+*
+* @example
+* // Bad...
+* if (foo) bar();
+*
+* // Good...
+* if (foo) {
+*   bar();
+* }
+*/
+rules[ 'curly' ] = 'error';
+
+/**
+* Requires a default case in switch statements.
+*
+* @name default-case
+* @memberof rules
+* @type {string}
+* @default 'error'
+* @see [default-case]{@link https://eslint.org/docs/rules/default-case}
+*
+* @example
+* // Bad...
+* switch (foo) {
+*   case 1:
+*     doSomething();
+* }
+*
+* // Good...
+* switch (foo) {
+*   case 1:
+*     doSomething();
+*     break;
+*   default:
+*     doNothing();
+* }
+*/
+rules[ 'default-case' ] = 'error';
+
+/**
+* Disables the rule for dot notation.
+*
+* @name dot-notation
+* @memberof rules
+* @type {string}
+* @default 'off'
+* @see [dot-notation]{@link https://eslint.org/docs/rules/dot-notation}
+*
+* @example
+* // Okay...
+* const x = foo[ 'bar' ];
+*/
+rules[ 'dot-notation' ] = 'off';
+
+/**
+* Enforces a newline at the end of files.
+*
+* @name eol-last
+* @memberof rules
+* @type {string}
+* @default 'error'
+* @see [eol-last]{@link https://eslint.org/docs/rules/eol-last}
+*
+* @example
+* // Bad...
+* const bad = 'no newline';
+* // EOF
+*
+* // Good...
+* const good = 'newline';
+* // EOF
+*/
+rules[ 'eol-last' ] = 'error';
+
+/**
+* Enforces the use of `===` and `!==` in place of `==` and `!=`.
+*
+* @name eqeqeq
+* @memberof rules
+* @type {Array}
+* @default [ 'error', 'always' ]
+* @see [eqeqeq]{@link https://eslint.org/docs/rules/eqeqeq}
+*
+* @example
+* // Bad...
+* if (x == y) {
+*   // ...
+* }
+*
+* // Good...
+* if (x === y) {
+*   // ...
+* }
+*/
+rules[ 'eqeqeq' ] = [ 'error', 'always' ];
+
+/**
+* Requires `for-in` loops to be filtered with an `if` statement.
+*
+* @name guard-for-in
+* @memberof rules
+* @type {string}
+* @default 'error'
+* @see [guard-for-in]{@link https://eslint.org/docs/rules/guard-for-in}
+*
+* @example
+* // Bad...
+* for (const key in object) {
+*   doSomething(object[key]);
+* }
+*
+* // Good...
+* for (const key in object) {
+*   if (Object.prototype.hasOwnProperty.call(object, key)) {
+*     doSomething(object[key]);
+*   }
+* }
+*/
+rules[ 'guard-for-in' ] = 'error';
+
+/**
+* Requires identifiers to match a specified regular expression.
+*
+* @name id-match
+* @memberof rules
+* @type {string}
+* @default 'error'
+* @see [id-match]{@link https://eslint.org/docs/rules/id-match}
+*
+* @example
+* // Bad...
+* const my_var = 'bad';
+*
+* // Good...
+* const myVar = 'good';
+*/
+rules[ 'id-match' ] = 'error';
```

Comment:
Does this need a regular expression?

Reply:
Yes, looks like it: https://eslint.org/docs/latest/rules/id-match

I am seeing we have this rule turned off in our JS `eslint` configuration. 

Shall we do the same here or try to e.g. enforce camelCase and CONSTANT_CASE?

Reply:
Do we include a rationale for why this is disabled in our JS ESLint configuration? Or any clues in our commit history?

Reply:
@kgryte I ended up disabling it after playing around with a few regular expressions since they flagged too many variables (we say in the JSDoc for the JS code that we don't want to impose draconian naming convention). Maybe we can revisit this though as I think it would be nice to e.g. rule out `snake_case`.

Reply:
@Planeshifter If you can, just create a TODO issue which discusses the current state of things (e.g., current practice, along with representative examples), and a proposed convention that we will enforce.

Reply:
Issue created on internal TODO tracker.